cpp/src/arrow/util/bpacking_avx512_icx.h

```diff
@@ -0,0 +1,1145 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```

Comment:
How was this code written? By hand? Was it generated? How is it supposed to be maintained?

Reply:
@pitrou Hey, thanks for your question. This code is written by hand now. I agree that AVX512 intrinsic is not obvious and not easy to understand. But it shows great performance improvement. For this case up to 55% performance improvement is observed on the Intel ICX platform. Regarding the maintainance issue, we can verify its function via unit tests, function tests, etc. If more functions/features are introduced in the future, we will add more test cases to ensure it works as designed.